### ISSUE TITLE
Contract Items update can take an extremely long time due to error throttling and lack of caching

### ISSUE THREAD

**Issue by elasota** [GoldenGnu/jeveassets#360]
Hey, I've been running into a problem of some sort that's been getting worse lately.  If I go to "Update" -> "Update..." then it will frequently get stuck or take an extremely long time on "Contract Items."

Right now it is bad enough that jeveassets' work threads are erroring out and it takes almost 15 minutes to do a full update because it's getting throttled by ESI:
```
WARN: 2022-12-01 15:54:23,566 [pool-43-thread-45] net.nikr.eve.jeveasset.io.esi.AbstractEsiGetter - Error limit reached waiting: 1s
WARN: 2022-12-01 15:54:24,576 [pool-43-thread-41] net.nikr.eve.jeveasset.io.shared.AbstractGetter - ESI Contract Items failed to update for:  > Successful Businessman ERROR: {"error":"Undefined 429 response. Original message: Too many errors.You have been temporarily throttled."}
WARN: 2022-12-01 15:54:24,576 [pool-43-thread-33] net.nikr.eve.jeveasset.io.shared.AbstractGetter - ESI Contract Items failed to update for:  > Successful Businessman ERROR: {"error":"This software has exceeded the error limit for ESI. If you are a user, please contact the maintainer of this software. If you are a developer/maintainer, please make a greater effort in the future to receive valid responses. For tips on how, come have a chat with us in #esi on tweetfleet slack. If you're not on tweetfleet slack yet, you can get an invite here -> https://www.fuzzwork.co.uk/tweetfleet-slack-invites/"}
```

This looks like a problem with jeveassets since I have my own asset tracker tool and it isn't getting stuck on contracts.  However, I would note some things:

- I have a large number of contracts since I run a business that's mainly conducted via public contracts and ESI will return all contracts going back for a year.  Currently logs show that jeveassets is trying to retrieve the contents of over 3000 contracts.
- jeveassets appears to query contract items for all contracts returned from the API, even though it already has the contract items available from previous queries.  This is very wasteful, and runs up a lot of errors since the contract items endpoints are kind of flaky (especially the corp one).  Contract items never change, so it should only be querying the items of contracts that it doesn't have already.

**Comment by GoldenGnu** [comment]
Thank you for your bug report.
A few version back I added updating from the public contract items endpoint, as it have detailed information about blueprints, that the corp/char endpoint doesn't (BPO/ME/TE), I guess the code isn't selective enough of what it can and can not update from the public endpoint. I will look into it.

**Comment by elasota** [comment]
FWIW most of those contracts are blueprint copies sold on public contracts.

Some more info from looking at the logs (I can send the log if you want):

There are a lot of "Contract not found" and "Contract not public" errors in the log that appear to all be linked to corporation contracts.  Not sure what's going on with the "Contract not found" errors, but the corp contracts endpoint returns the availability of the contract, so it shouldn't be querying the public contracts endpoint for non-public contracts.  I'm only seeing this for corp contracts.

Second, most of the errors that are showing stack traces are showing `net.troja.eve.esi.api.ContractsApi.getCharactersCharacterIdContractsContractIdItemsWithHttpInfo` which is using the character contract items API... and the blueprint ME/TE/Runs aren't showing in the Tools -> Business -> Contracts table anyway.  So, this doesn't look like a problem of it using the public contracts endpoint.

**Comment by GoldenGnu** [comment]
Maybe you debug it yourself and submit a PR?

**Comment by GoldenGnu** [comment]
I made a test build, that I hope fixes this issue (or at least improve it).

**Download:** [jEveAssets Test Build](eve.nikr.net/jeveassets/jeveassets-7.4.1-contracts1.zip)

**Info:** Please read the [dev build article on the wiki](https://wiki.jeveassets.org/devbuild) before downloading (or unwanted things could happen)

**Feedback:** Let me know if it resolve or improve the issue or not.

Just a bit of extra information:
Contract items endpoint are rate limited by monolith, hence why it can take a long time to update a lot of contracts even without errors.

**Comment by elasota** [comment]
Sorry for the delay, I tried the test build and it seems much faster.

**Comment by GoldenGnu** [comment]
Thank you very much for testing the dev build and verifying the fix! 👍

**Comment by GoldenGnu** [comment]
Released in 7.5.0
